bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -2,1383 +2,708 @@
-        shared_controller_state: Arc<SharedVmState>,
-        vcpu_tasks: C,
-        log: Logger,
-        api_state_tx: tokio::sync::watch::Sender<ApiMonitoredState>,
+        log: slog::Logger,
+        auto_start: InstanceAutoStart,
     ) -> Self {
         Self {
-            runtime_hdl,
-            controller,
-            shared_state: shared_controller_state,
-            vcpu_tasks,
-            log,
-            state_gen: 0,
-            paused: false,
-            api_state_tx,
+            inner: Mutex::new(InputQueueInner::new(log, auto_start)),
+            notify: Notify::new(),
         }
     }
 
-    /// Yields the current externally-visible instance state.
-    fn get_instance_state(&self) -> ApiInstanceState {
-        self.api_state_tx.borrow().state
-    }
-
-    /// Retrieves the most recently published migration state from the external
-    /// migration state channel.
+    /// Waits for an event to arrive on the input queue and returns it for
+    /// processing.
     ///
-    /// This function does not return the borrowed monitor, so the state may
-    /// change again as soon as this function returns.
-    fn get_migration_status(&self) -> ApiMigrateStatusResponse {
-        self.api_state_tx.borrow().migration.clone()
-    }
+    /// External requests and guest events are stored in separate queues. If
+    /// both queues have events when this routine is called, the guest event
+    /// queue takes precedence.
+    ///
+    /// # Synchronization
+    ///
+    /// This routine assumes that it is only ever called by one task (the state
+    /// driver). If multiple threads call this routine simultaneously, they may
+    /// miss wakeups and not return when new events are pushed to the queue or
+    /// cause a panic (see below).
+    async fn wait_for_next_event(&self) -> InputQueueEvent {
+        loop {
+            {
+                let mut guard = self.inner.lock().unwrap();
+                if let Some(guest_event) = guard.guest_events.pop_front() {
+                    return InputQueueEvent::GuestEvent(guest_event);
+                } else if let Some(req) = guard.external_requests.pop_front() {
+                    return InputQueueEvent::ExternalRequest(req);
+                }
+            }
 
-    /// Sets the published instance and/or migration state and increases the
-    /// state generation number.
-    fn set_published_state(&mut self, state: PublishedState) {
-        let (instance_state, migration_state) = match state {
-            PublishedState::Instance(i) => (Some(i), None),
-            PublishedState::Migration(m) => (None, Some(m)),
-            PublishedState::Complete(i, m) => (Some(i), Some(m)),
-        };
+            // Notifiers in this module must use `notify_one` so that their
+            // notifications will not be lost if they arrive after this routine
+            // checks the queues but before it actually polls the notify.
+            self.notify.notified().await;
```

Comment:
I think we should be using [`Notified::enable`](https://docs.rs/tokio/latest/tokio/sync/futures/struct.Notified.html#method.enable) here, to subscribe to a potential wakeup _before_ checking the event queues. This way, if there's a race condition where new elements are pushed to the queue but we have not yet started to wait for a notification, we won't miss the wakeup.

I think the correct code would look something like this:

```rust
        let notified = self.notify.notified();
        tokio::pin!(notified);
        loop {
            // Pre-subscribe to a wakeup from `self.notify` so that we don't
            // miss any notification that is sent while checking the event queues.
            notify.as_mut().enable();

            // Check whether there are any messages in either queue.
            {
                let mut guard = self.inner.lock().unwrap();
                if let Some(guest_event) = guard.guest_events.pop_front() {
                    return InputQueueEvent::GuestEvent(guest_event);
                } else if let Some(req) = guard.external_requests.pop_front() {
                    return InputQueueEvent::ExternalRequest(req);
                }
            }
            
            // Await a notification if the queues were empty.
            notified.as_mut().await;
            // Reset the `notified` future in case we must await an additional
            // notification.
            notified.set(self.notify.notified()
        }
```